lib/node/server.ml

```diff
@@ -1,51 +1,125 @@
+    | Post ->
+      if not (Disseminator.seen !node.disseminator message) then (
+        let%lwt () =
+          log node
+            (Printf.sprintf "%s:%d : Processing post %s from %s:%d\n"
+               !node.address.address !node.address.port
+               (Message.hash_of message) message.sender.address
+               message.sender.port) in
+        let _ = msg_handler message in
+        let%lwt () = log node "Adding message to broadcast queue\n" in
+        Client.post node message;
+        Lwt.return ())
+      else
+        log node
+          (Printf.sprintf "Got post %s from %s:%d but saw it already\n"
+             (Message.hash_of message) message.sender.address
+             message.sender.port)
+    | _ ->
+      let _ = msg_handler message in
+      Lwt.return () in
+  Lwt.return ()
+
+(** Log some initial information at the beginning of a server iteration.
+    See comments for descriptions regarding what is actually being logged. *)
+let print_logs node =
+  (* Check that the server is in fact running *)
+  let%lwt () = log node "Running server\n" in
+  (* Check which posts the node has seen so far *)
+  let%lwt () =
+    log node
+      (Printf.sprintf "Seen: %s\n"
+         (Disseminator.all_seen !node.disseminator |> String.concat " ; "))
+  in
+  (* Check who the current peers of the node are *)
+  let%lwt () =
+    !node.peers
+    |> Base.Hashtbl.keys
+    |> List.map (fun Address.{ port; _ } -> string_of_int port)
+    |> String.concat " ; "
+    |> Printf.sprintf "Peers: %s\n"
+    |> log node in
+  (* Check which posts are currently being disseminated by the node *)
+  if List.length (Disseminator.broadcast_queue !node.disseminator) > 0 then
     let%lwt () =
-      match request with
-      | Some request ->
-        let response =
-          request |> msg_handler |> Client.create_response node request in
-        let%lwt () = Client.send_to node response in
-        Lwt.return ()
-      | None -> Lwt.return () in
-
-    server () in
-  Lwt.async server
+      !node.disseminator
+      |> Disseminator.broadcast_queue
+      |> List.map Message.hash_of
+      |> String.concat " "
+      |> Printf.sprintf "Broadcast Queue: %s\n"
+      |> log node in
+    Lwt.return ()
+  else
+    Lwt.return ()
+
+(* Sever procedure:
+   0. Log pertinent information about the current node.
+   1. Start a new thread for handling any incoming message.
+   2. Run the failure detector.
+   3. Run the disseminator, this includes actually sending messages to be
+      disseminated across the network.
+   4. Wait 0.001 seconds before restarting the procedure. *)
+let rec run node preprocessor msg_handler =
+  (* Step 0 *)
+  let%lwt () = print_logs node in
+  (* Step 1 *)
+  let _ = process_message node preprocessor msg_handler in
+  (* Step 2 *)
+  let%lwt () = Failure_detector.failure_detection node in
+  (* Step 3 *)
+  let%lwt () = Networking.disseminate node in
+  (* Step 4 *)
+  let%lwt () = Lwt_unix.sleep 0.001 in
```

Comment:
Hum, not fond of this, but can't find an other way to achieve this.
Aren't the execution of `failure_detection` and `disseminate` enough to "lose" time?